### PR TITLE
Slides to Videos

### PR DESCRIPTION
**Pull request by hexylena**
Thanks to the work by @Delphine-L and others (see [https://github.com/jxtx/covid-lessons](jxtx/covid-lessons)) we've implemented automatic slides-to-video conversion. This is an opt-in feature for slide authors, enabled by adding `video: true` to the metadata.

[image](https://galaxy-training.s3.amazonaws.com/videos/topics/admin/tutorials/ansible/slides.mp4)

## [See the Demo](https://galaxy-training.s3.amazonaws.com/videos/topics/admin/tutorials/ansible/slides.mp4)

 There is audio and a subtitle track embedded for players which are aware.

Additionally this PR replaces the slow decktape/chrome PDF build process with one based on wkhtmltopdf which is significantly faster and parallelisable in a way that decktape/chrome would never be. These are additionally converted to proper make targets so when you build videos, only the required slide decks need to be built.

Chrome/Decktape | wkhtmltopdf
----  | --- 
image | image

(If we're being honest, they're both bad in their own ways buuuut whatever. at least the "Chat on Gitter" is gone from every single page (as chrome scrolled.) and the useless localhost URL is gone too.)

TODOs:

- [x] We will need to outline the criteria for speaker notes, as they are different than standard speaker notes a bit!
- [x] Embed / link videos in the interface somewhere
- [x] Request polly permission on our IAM key.

### CODE REVIEW

**Comment by hexylena** [comment]
Videos are linked from the topic pages:

image

And are displayed in one long list:

image

Subtitles will **NOT** work until https://github.com/galaxyproject/infrastructure-playbook/pull/30 is merged.

**Comment by hexylena** [comment]
# Hosting

We're starting by just dumping on the S3 bucket, this is easy and names are predictable.

Going forward:

- We will look into uploading into vimeo or youtube or similar (opinions welcome.)
- We'll have to have a script to upload the video, and then open a PR to commit the video ID back into the GTN.

# Building

We'll will start by just building them as needed. 

Going forward:

- We'll setup a cron job to check if there were any changes since the last time the cron job ran (ugh, not super fun.)
- And if any of the slides were touched since then, then build those videos and upload, and commit the new IDs.

**Review by abretaud — COMMENTED** [review]
Cool work :+1: !
While preparing for a session, I've tested the new pdf build process, and had a few little problems:
- [ ] I get wkhtmltopdf errors (`Exit with code 1 due to network error: ContentNotFoundError`) for tutorials having embedded vimeo videos. I guess it could be replaced by an url in the pdf?
- [x] the slides/*html files were not built, I had to change the Makefile (see suggestions)
- [ ] everything in italic did not appear in the slides pdf, but not sure if it's specific to my laptop. I had to add this css to show them back: ```em {
  font-family: 'Ubuntu Mono';
}```
- [ ] for some reason, for slides, wkhtmltopdf installed with conda produced a single page with only the background and no content. Installing fedora's version fixed it. Maybe also something to do with my laptop?
- [x] also a missing quote in the css (suggestion too)

File: assets/css/slides.css [hunk 1]
```diff
@@ -77,7 +76,7 @@ h3 {
 
 .remark-code,
 .remark-inline-code {
-  font-family: 'Ubuntu Mono', 'Courier';
+  font-family: 'Courier', monospace;
```
[the remark this answers is not in the document]
Comment (nsoranzo) [reply assets/css/slides.css:79]:
```suggestion
  font-family: 'Courier', 'monospace';
```

Comment (hexylena) [reply assets/css/slides.css:79]: This should not be quoted, monospace there is a special value indicating the browser should choose a built-in monospace font. https://developer.mozilla.org/en-US/docs/Web/CSS/font-family

Comment (hexylena) [reply assets/css/slides.css:79]: "Should" is an unnecessarily strong word from me. Let me rephrase: I have never seen the generic font families quoted. I do not think it is breaks anything, but it isn't necessary.

File: Makefile [hunk 2]
```diff
@@ -188,33 +178,39 @@ check-links-gh-pages:  ## validate HTML on gh-pages branch (for daily cron job)
...
+	if ! grep 'http-equiv="refresh"' $< --quiet; then \
+		$(ACTIVATE_ENV) && \
+		sed "s|/training-material/|$(shell pwd)/_site/training-material/|g" $< | \
+		sed "s|<head>|<head><base href=\"file://$(shell pwd)/$(<:_site/training/material%=%)\">|" | \
+		wkhtmltopdf \
+		    --enable-javascript --javascript-delay 1000 \
+			- $@; \
+	fi
+
+_site/%/slides.pdf: _site/%/slides.html
```
[the remark this answers is not in the document]
Comment (hexylena) [reply Makefile:197]: Good points!

**Comment by hexylena** [comment]
>  for tutorials having embedded vimeo videos. I guess it could be replaced by an url in the pdf?

Ahh interesting. There's no easy way to work around this. Given that only a few trainings have these (and we generally try and get rid of them whenever they occur), I think it's maybe ok if we just ignore these failures?

> everything in italic did not appear in the slides pdf

So that seems to be because the files specify remote fonts. I tried adding a `,serif;` to the droid sans usage but that doesn't help. I guess it needs to point to a local ttf, or we need an override for the slides. Easy enough to do.

https://github.com/wkhtmltopdf/wkhtmltopdf/issues/3631

> wkhtmltopdf installed with conda produced a single page with only the background and no content. Installing fedora's version fixed it

I don't use the conda package either to be honest, I use my system's. I added it because conda is used for other things here but never tested it. Not sure what to do about that, but since the PDF output is not commonly used, I think it's lower priority maybe?

**Comment by abretaud** [comment]
> > for tutorials having embedded vimeo videos. I guess it could be replaced by an url in the pdf?
> 
> Ahh interesting. There's no easy way to work around this. Given that only a few trainings have these (and we generally try and get rid of them whenever they occur), I think it's maybe ok if we just ignore these failures?

Sure, I'm not using these tutorials' pdf actually, no problem to ignore the errors (though, maybe the makefile stops at the first error? I don't remember)

> 
> > everything in italic did not appear in the slides pdf
> 
> So that seems to be because the files specify remote fonts. I tried adding a `,serif;` to the droid sans usage but that doesn't help. I guess it needs to point to a local ttf, or we need an override for the slides. Easy enough to do.
> 
> [wkhtmltopdf/wkhtmltopdf#3631](https://github.com/wkhtmltopdf/wkhtmltopdf/issues/3631)

:+1: 

> > wkhtmltopdf installed with conda produced a single page with only the background and no content. Installing fedora's version fixed it
> 
> I don't use the conda package either to be honest, I use my system's. I added it because conda is used for other things here but never tested it. Not sure what to do about that, but since the PDF output is not commonly used, I think it's lower priority maybe?

Yep low priority for me too

**Comment by hexylena** [comment]
Sounds good! Thank you for testing, super appreciate all of the reports so we at least know under which circumstances it might have difficulties :)

**Review by shiltemann — APPROVED** [review]

**Comment by shiltemann** [comment]
:tada: so awesome! thanks!

**Comment by bgruening** [comment]
Lovely!!! :heart:

**Comment by martenson** [comment]
Neither [conda-forge](https://anaconda.org/conda-forge/wkhtmltopdf/files) or [bioconda](https://anaconda.org/bioconda/wkhtmltopdf/files) have binaries of wkhtmltopdf for MacOS which makes apple folks unable to build training tutorials easily.

**Comment by hexylena** [comment]
The PDFs were not heavily used, to our knowledge? We never produced them automatically within the project.

This does not affect anyone building the rest of the tutorials. This might prevent them from building the videos, but that's also not possible without AWS credentials, which not everyone has, and we just want it to be automatic for users, not something they have to build locally to test first?

**Comment by martenson** [comment]
I might be doing things wrong, it's been some time I ran training locally:

```
make install
if /Users/marten/miniconda3/condabin/conda env list | grep '^galaxy_training_material'; then \
	    /Users/marten/miniconda3/condabin/conda env update -f environment.yml; \
	else \
	    /Users/marten/miniconda3/condabin/conda env create -f environment.yml; \
	fi
galaxy_training_material  *  /Users/marten/miniconda3/envs/galaxy_training_material
Collecting package metadata (repodata.json): done
Solving environment: failed

ResolvePackageNotFound:
  - wkhtmltopdf

make: *** [create-env] Error 1
```

**Comment by hexylena** [comment]
Oh! Ok, no, you're right, I added that to the main environment when it isn't needed. Thanks for the explanation, I'll remove that shortly.
